File: src/freight_fate/online_activation.py
"""Device-code activation: replaces the clipboard-paste credential setup.

The old flow made a player copy a Driver ID and a token from the Orinks
driver setup page and paste both into the game (see
:mod:`freight_fate.online_presence`'s :class:`OnlineIdentity`). That is
error-prone for a screen reader user working across two applications with no
shared clipboard review. This module instead drives the OAuth-device-code
style exchange already live on orinks.net: the game asks for a short code,
speaks it, the player types that code into any browser (any device) and
signs in there, and the game polls until the site says the code was claimed.

This is the *only* place that knows about the two activation endpoints. It
deliberately reuses :data:`freight_fate.online_presence.Transport`,
:func:`freight_fate.online_presence._http_json`, and
:func:`freight_fate.online_presence.base_url` rather than forking them, so
the game has exactly one HTTP client and exactly one
``FREIGHT_FATE_ONLINE_URL`` override for every Orinks endpoint.

Both request/response shapes below are an existing, already-deployed server
contract -- do not change field names or status-code meanings here without
updating the server first.
"""

from __future__ import annotations

import hashlib
import logging
import platform
import time
import urllib.error
import uuid
from dataclasses import dataclass, field

from .online_presence import Transport, _http_json, base_url

log = logging.getLogger(__name__)


def machine_key() -> str:
    """An opaque, stable name for this computer, for the computer list.

    orinks.net caps an account at ten computers and used to count
    ACTIVATIONS: every freshly unzipped build that connected took another
    slot, so a tester who takes a build a week filled the list with one PC
    and was locked out (armstrong445, 2026-08-15). Sending this lets the
    server replace that computer's entry instead of adding one.

    Hashed, and never the hostname itself: the server only ever compares it
    for equality, so there is no reason to hand over a name that might be a
    person's. Derived from the machine's network identity, which is stable
    across a reinstall and across unzipping the game somewhere else -- the
    two things that were minting duplicate rows.

    This is NOT identity and never authenticates anything: it is a hint the
    player triggers by deliberately connecting, and a wrong or missing one
    costs only a duplicate row.
    """
    seed = f"{uuid.getnode():x}:{platform.node()}"
    return hashlib.sha256(seed.encode("utf-8", "replace")).hexdigest()[:32]


__all__ = [
    "Activation",
    "machine_key",
    "PollResult",
    "spell_code",
    "start_activation",
    "poll_activation",
    "base_url",
]

# NATO phonetics for every letter the activation alphabet could ever contain.
# The alphabet itself (ABCDEFGHJKMNPQRTUVWXY346789, defined server-side)
# excludes O I L S Z 0 1 2 5 specifically so no two of these words are ever
# close enough to be confused for each other over a screen reader -- keeping
# the unused letters here too costs nothing and means this table never has
# to change if the server's alphabet ever does.
_PHONETIC = {
    "A": "Alpha",
    "B": "Bravo",
    "C": "Charlie",
    "D": "Delta",
    "E": "Echo",
    "F": "Foxtrot",
    "G": "Golf",
    "H": "Hotel",
    "I": "India",
    "J": "Juliett",
    "K": "Kilo",
    "L": "Lima",
    "M": "Mike",
    "N": "November",
    "O": "Oscar",
    "P": "Papa",
    "Q": "Quebec",
    "R": "Romeo",
    "S": "Sierra",
    "T": "Tango",
    "U": "Uniform",
    "V": "Victor",
    "W": "Whiskey",
    "X": "Xray",
    "Y": "Yankee",
    "Z": "Zulu",
}

_DIGITS = {
    "0": "zero",
    "1": "one",
    "2": "two",
    "3": "three",
    "4": "four",
    "5": "five",
    "6": "six",
    "7": "seven",
    "8": "eight",
    "9": "nine",
}


def spell_code(code: str) -> str:
    """Spell an activation code letter-by-letter for a screen reader.

    The game has no review cursor, so a player cannot step through a spoken
    string character by character the way they can in a browser -- speaking
    ``WKQR-3468`` once, as a word, gives them nothing to transcribe. This
    returns NATO phonetics for letters and plain words for digits,
    comma-separated (so a screen reader pauses between entries), with the
    dash spoken too so a player copying the code by ear knows it belongs in
    the string. Works on a code with or without the dash.
    """
    words = []
    for ch in code:
        if ch == "-":
            words.append("dash")
            continue
        upper = ch.upper()
        if upper in _PHONETIC:
            words.append(_PHONETIC[upper])
        elif ch in _DIGITS:
            words.append(_DIGITS[ch])
        else:
            # Defensive only: the server-issued alphabet never produces
            # anything else, but an unrecognised character still gets read
            # out verbatim instead of silently vanishing from the spelling.
            words.append(ch)
    return ", ".join(words)


@dataclass
class Activation:
    """An in-progress device-code activation.

    ``device_code`` is the polling secret -- it is bound to this device and
    never shown to the player, so it must never be logged or included in any
    spoken or transcript-bound string. It is ``repr=False`` so that invariant
    is structural rather than a rule every future caller has to remember: a
    stray ``log.warning("... %r", activation)`` would otherwise write the
    secret straight into the session log with nothing failing.
    ``user_code`` is the short code the
    player reads back and types into a browser. ``expires_at`` is a
    ``time.time()``-based deadline (the server gives a relative
    ``expires_in`` in seconds; this module resolves it to an absolute time
    once, at start, so a caller checking it later doesn't need to remember
    when the request was made).
    """

    device_code: str = field(repr=False)
    user_code: str
    verification_uri: str
    verification_uri_complete: str
    expires_at: float
    interval: float


@dataclass
class PollResult:
    """One poll's outcome.

    ``status`` is one of ``"pending"`` (keep waiting), ``"ready"`` (claimed
    -- ``driver_id``, ``token`` and ``display_name`` are all set),
    ``"expired"`` (the code timed out, or was over an account's device cap;
    either way the fix is the same: start over with a new code),
    ``"retry"`` (a transient failure -- network trouble, an HTTP status other
    than 400/410, a reply the caller could not parse, or a "ready" body with
    no ``driver_id``/``token`` in it -- that a caller
    should just poll again on the next tick, exactly like "pending"), or
    ``"error"`` (HTTP 400 specifically: the server rejected the stored
    device_code as malformed. Retrying the same code can never fix this, so
    it must not be presented to the player as "expired" or as something a
    moment's wait will resolve, the way "retry" is).
    """

    status: str
    driver_id: str | None = None
    token: str | None = None
    display_name: str | None = None


def start_activation(*, transport: Transport = _http_json) -> Activation | None:
    """Ask orinks.net to mint a new device code, or None if it could not.

    None covers every failure the player can't do anything about but wait
    and retry: rate limiting (429), the endpoint being down (503), a
    malformed reply, or a network error. The caller (the setup menu) is
    expected to show one generic "couldn't reach Orinks, try again" message
    for all of them -- there is nothing actionable to say differently.
    """
    try:
        # An older server ignores an unknown field, so this is safe to send
        # before the site half is everywhere.
        reply = transport(
            f"{base_url()}/api/freight-fate/activate/start", {"machine_key": machine_key()}, {}
        )
    except Exception as e:
        log.warning("Activation start failed: %s", e)
        return None
    try:
        return Activation(
            device_code=reply["device_code"],
            user_code=reply["user_code"],
            verification_uri=reply["verification_uri"],
            verification_uri_complete=reply["verification_uri_complete"],
            expires_at=time.time() + float(reply["expires_in"]),
            interval=float(reply["interval"]),
        )
    except (KeyError, TypeError, ValueError) as e:
        log.warning("Activation start returned a malformed reply: %s", e)
        return None


def poll_activation(activation: Activation, *, transport: Transport = _http_json) -> PollResult:
    """Check whether ``activation``'s code has been claimed yet.

    Runs on a timer while the player waits at the setup screen, so no
    exception may ever escape to the caller -- a transient network blip must
    not crash the menu. Never logs or otherwise surfaces
    ``activation.device_code``.

    "error" is reserved for HTTP 400 alone. Everything else this function
    cannot make sense of -- a dropped connection, a timeout, an SSL error, a
    5xx, an unexpected status code, a 200 body that is not even a mapping --
    comes back as "retry" instead, so a caller (``OnlineSetupState``) can
    just poll again on the next tick rather than forcing the player through
    a fresh activation code for what is very likely a momentary blip.
    """
    try:
        reply = transport(
            f"{base_url()}/api/freight-fate/activate/poll",
            {"device_code": activation.device_code},
            {},
        )
    except urllib.error.HTTPError as e:
        if e.code == 410:
            # Covers both a timed-out code and an over-cap redeem -- the
            # player learns the real reason (too many computers on the
            # account) in the browser at claim time, so the game just
            # treats both as "get a new code".
            return PollResult(status="expired")
        if e.code == 400:
            # Malformed device_code is not "expired": retrying the same code
            # can never fix it, so it must surface as a distinct status
            # rather than telling the player to just wait it out.
            log.warning("Activation poll rejected the stored device_code as malformed")
            return PollResult(status="error")
        # Any other HTTP status (429, 5xx, ...) is treated as transient: none
        # of them mean the code itself is bad, so the same code can just be
        # polled again.
        log.warning("Activation poll failed: HTTP %s", e.code)
        return PollResult(status="retry")
    except Exception as e:
        # URLError, socket timeouts, SSL errors, and anything else that is
        # not an HTTPError are all connectivity trouble, not a verdict on the
        # code -- transient, same as an HTTP 5xx above.
        log.warning("Activation poll failed: %s", e)
        return PollResult(status="retry")

    # A 200 body that isn't even a mapping (None, a list, a bare string --
    # anything a broken deploy or a middlebox could hand back) must land here
    # too, not raise: this whole block, not just the HTTPError branch above,
    # is the "never let an exception escape" guarantee the docstring makes.
    try:
        status = reply.get("status")
        if status == "ready":
            driver_id = reply.get("driver_id")
            token = reply.get("token")
            if not driver_id or not token:
                # A "ready" that carries no identity is not a claim the caller
                # can act on: saving it would write a null identity, tell the
                # player "Connected to orinks.net", and then quietly send
                # "Bearer None" on every heartbeat until the next launch drops
                # it. Treat it the same as any other body this module cannot
                # make sense of -- transient, keep polling -- so a broken
                # deploy or a body-rewriting middlebox resolves itself once
                # the real claim lands, and expiry (not a wrong success) is
                # the worst outcome.
                log.warning("Activation poll returned a ready reply with no identity")
                return PollResult(status="retry")
            return PollResult(
                status="ready",
                driver_id=driver_id,
                token=token,
                display_name=reply.get("display_name"),
            )
        if status == "pending":
            return PollResult(status="pending")
    except AttributeError:
        log.warning("Activation poll returned a non-object reply: %r", reply)
        return PollResult(status="retry")

    # An unrecognised 200 body (a mapping missing "status", or an unknown
    # value) is most plausibly a transient server or proxy hiccup rather than
    # evidence about the code itself, so it gets the same "poll again" status
    # as a network blip -- not the terminal "error" that only HTTP 400 gets.
    log.warning("Activation poll returned an unexpected status: %r", status)
    return PollResult(status="retry")

File: crates/freight-fate/tests/it/online_activation.rs
//! Port of `tests/test_online_activation.py`: the device-code activation
//! flow (replaces clipboard paste).
//!
//! A fake transport keeps every test free of real sockets, matching the
//! pattern in `online_presence.rs`. The contract points that matter most:
//! 410 means "expired, get a new code" while 400 means "this device_code is
//! malformed and retrying it never helps" -- "error" is reserved for exactly
//! that 400 case, and everything else unrecognised (other HTTP statuses,
//! network trouble, an unparseable 200 body) is "retry" instead, so the two
//! must never collapse to the same status; an over-cap redeem also answers
//! 410, because the real reason lives on the website at claim time; and
//! `display_name` must survive a ready poll uncorrupted, because it is the
//! player's only signal that the code was claimed on the wrong account.

use std::sync::Mutex;

use serde_json::{json, Value};

use freight_fate::net::testing::ClosureTransport;
use freight_fate::net::{NetError, Transport};
use freight_fate::online_activation::{
    self, base_url, machine_key, poll_activation, spell_code, start_activation, Activation,
};

fn an_activation() -> Activation {
    Activation {
        device_code: "a".repeat(64),
        user_code: "WKQR-3468".to_string(),
        verification_uri: "https://orinks.net/activate".to_string(),
        verification_uri_complete: "https://orinks.net/activate?code=WKQR-3468".to_string(),
        expires_at: 0.0,
        interval: 3.0,
    }
}

fn replying(body: Value) -> impl Transport {
    ClosureTransport(
        move |_url: &str, _p: Option<&Value>, _h: &[(String, String)], _m: Option<&str>| {
            Ok(body.clone())
        },
    )
}

fn poll_raising(error: NetError) -> online_activation::PollResult {
    let transport = ClosureTransport(
        move |_url: &str, _p: Option<&Value>, _h: &[(String, String)], _m: Option<&str>| {
            Err(error.clone())
        },
    );
    poll_activation(&an_activation(), &transport)
}

fn data_dir() -> tempfile::TempDir {
    tempfile::tempdir().unwrap()
}

// -- spell_code ----------------------------------------------------------------

#[test]
fn test_spell_code_uses_phonetics_and_speaks_the_dash() {
    assert_eq!(
        spell_code("WKQR-3468"),
        "Whiskey, Kilo, Quebec, Romeo, dash, three, four, six, eight"
    );
}

#[test]
fn test_spell_code_accepts_an_undashed_code() {
    assert!(spell_code("WKQR3468").starts_with("Whiskey, Kilo"));
}

#[test]
fn test_spell_code_accepts_a_lowercase_code() {
    assert_eq!(
        spell_code("wkqr-3468"),
        "Whiskey, Kilo, Quebec, Romeo, dash, three, four, six, eight"
    );
}

#[test]
fn test_spell_code_covers_the_whole_activation_alphabet() {
    // ABCDEFGHJKMNPQRTUVWXY346789 -- deliberately excludes O I L S Z 0 1 2 5,
    // chosen so no two phonetic words could ever be confused for each other.
    let alphabet = "ABCDEFGHJKMNPQRTUVWXY346789";
    let spelled = spell_code(alphabet);
    let words: Vec<&str> = spelled.split(", ").collect();
    assert_eq!(words.len(), alphabet.len());
    let unique: std::collections::HashSet<&str> = words.iter().copied().collect();
    assert_eq!(unique.len(), words.len());
    assert!(words.iter().all(|w| !w.is_empty()));
}

// -- the device_code never leaves this module ----------------------------------

#[test]
fn test_activation_repr_never_carries_the_device_code() {
    // The polling secret must never reach a log line or the session
    // transcript. Keeping it out by convention is not enough: a Debug dump
    // is what a stray `log::warn!("... {:?}", activation)` would print, and
    // nothing would fail. Leaving it out of Debug makes the invariant
    // structural.
    let mut activation = an_activation();
    activation.device_code = format!("s3cret{}", "a".repeat(58));

    let text = format!("{activation:?}");

    assert!(!text.contains(&activation.device_code));
    assert!(!text.contains("s3cret"));
    // The player-facing code is not a secret and is still there, so a repr
    // remains useful for diagnosing which activation is in play.
    assert!(text.contains("WKQR-3468"));
}

// -- start_activation ------------------------------------------------------------

fn start_reply() -> Value {
    json!({
        "device_code": "a".repeat(64),
        "user_code": "WKQR-3468",
        "verification_uri": "https://orinks.net/activate",
        "verification_uri_complete": "https://orinks.net/activate?code=WKQR-3468",
        "expires_in": 600,
        "interval": 3,
    })
}

#[test]
fn test_start_returns_an_activation() {
    let seen = Mutex::new(String::new());
    let transport = ClosureTransport(
        |url: &str, _p: Option<&Value>, _h: &[(String, String)], _m: Option<&str>| {
            *seen.lock().unwrap() = url.to_string();
            Ok(start_reply())
        },
    );
    let dir = data_dir();
    let activation = start_activation(&transport, dir.path()).expect("an activation");
    assert!(seen
        .lock()
        .unwrap()
        .ends_with("/api/freight-fate/activate/start"));
    assert_eq!(activation.user_code, "WKQR-3468");
    assert_eq!(activation.interval, 3.0);
    assert!(activation.expires_at > 0.0);
}

#[test]
fn test_start_names_this_computer_so_the_list_does_not_fill_up() {
    // orinks.net capped an account at ten computers and counted activations,
    // so every unzipped build took another slot and a tester filled the list with
    // one PC (armstrong445, 2026-08-15). The server replaces a computer's entry
    // when the game names it; naming it is this call's job.
    let sent = Mutex::new(Value::Null);
    let transport = ClosureTransport(
        |_url: &str, payload: Option<&Value>, _h: &[(String, String)], _m: Option<&str>| {
            *sent.lock().unwrap() = payload.cloned().unwrap_or(Value::Null);
            Ok(start_reply())
        },
    );
    let dir = data_dir();
    assert!(start_activation(&transport, dir.path()).is_some());
    let key = sent.lock().unwrap()["machine_key"]
        .as_str()
        .unwrap()
        .to_string();
    // Opaque and bounded: the server compares it and nothing else, so it must
    // never carry a hostname a person could be named in.
    assert_eq!(key.len(), 32, "{key}");
    assert!(key.chars().all(|c| "0123456789abcdef".contains(c)), "{key}");
    let host = online_activation::hostname().to_lowercase();
    if !host.is_empty() {
        assert!(!key.to_lowercase().contains(&host));
    }
    // The same computer must answer the same, or the row it replaces is never
    // found and the list fills up exactly as before.
    assert_eq!(key, machine_key(dir.path()));
}

#[test]
fn test_start_returns_none_on_rate_limit_or_unavailable() {
    let dir = data_dir();
    for code in [429u16, 503] {
        let transport = ClosureTransport(
            move |_url: &str, _p: Option<&Value>, _h: &[(String, String)], _m: Option<&str>| {
                Err(NetError::http(code))
            },
        );
        assert!(start_activation(&transport, dir.path()).is_none());
    }
}

#[test]
fn test_start_returns_none_on_malformed_reply() {
    // A 200 missing an expected field must not raise into the caller.
    let dir = data_dir();
    let transport = replying(json!({"device_code": "a".repeat(64)})); // missing user_code and friends
    assert!(start_activation(&transport, dir.path()).is_none());
}

#[test]
fn test_start_never_raises_on_network_trouble() {
    let dir = data_dir();
    let transport = ClosureTransport(
        |_url: &str, _p: Option<&Value>, _h: &[(String, String)], _m: Option<&str>| {
            Err(NetError::other("OSError", "no route to host"))
        },
    );
    assert!(start_activation(&transport, dir.path()).is_none());
}

// -- poll_activation ---------------------------------------------------------------

#[test]
fn test_poll_ready_carries_the_display_name() {
    // The display name is the player's only signal that someone else claimed
    // their code -- the game speaks it, so it must survive the poll.
    let transport = ClosureTransport(
        |url: &str, payload: Option<&Value>, _h: &[(String, String)], _m: Option<&str>| {
            assert!(url.ends_with("/api/freight-fate/activate/poll"));
            assert_eq!(
                payload.cloned().unwrap(),
                json!({"device_code": an_activation().device_code})
            );
            Ok(json!({
                "status": "ready",
                "driver_id": "rig-hauler",
                "token": format!("ffd_{}", "b".repeat(64)),
                "display_name": "Rig Hauler",
            }))
        },
    );
    let result = poll_activation(&an_activation(), &transport);
    assert_eq!(result.status, "ready");
    assert_eq!(result.driver_id.as_deref(), Some("rig-hauler"));
    assert_eq!(result.token, Some(format!("ffd_{}", "b".repeat(64))));
    assert_eq!(result.display_name.as_deref(), Some("Rig Hauler"));
}

#[test]
fn test_poll_ready_without_an_identity_is_a_retry_not_a_claim() {
    // A "ready" body with no driver_id or token is not something the caller
    // can act on. Trusting it would save a null identity and tell the player
    // "Connected to orinks.net" while every later heartbeat sent no token at
    // all -- a silent, unactionable failure. "retry" instead of "error" because
    // a broken deploy or a rewriting middlebox is transient from the game's
    // side: polling should keep going until the code really expires.
    let token = format!("ffd_{}", "b".repeat(64));
    let bodies = [
        json!({"status": "ready"}),
        json!({"status": "ready", "driver_id": "rig-hauler"}),
        json!({"status": "ready", "token": token}),
        json!({"status": "ready", "driver_id": "", "token": token}),
        json!({"status": "ready", "driver_id": "rig-hauler", "token": ""}),
        json!({"status": "ready", "driver_id": null, "token": null, "display_name": "Rig Hauler"}),
    ];
    for body in bodies {
        let result = poll_activation(&an_activation(), &replying(body.clone()));
        assert_eq!(result.status, "retry", "{body}");
        assert!(result.driver_id.is_none());
        assert!(result.token.is_none());
    }
}

#[test]
fn test_poll_pending_carries_no_identity() {
    let result = poll_activation(&an_activation(), &replying(json!({"status": "pending"})));
    assert_eq!(result.status, "pending");
    assert!(result.driver_id.is_none());
    assert!(result.token.is_none());
    assert!(result.display_name.is_none());
}

#[test]
fn test_poll_maps_410_to_expired_and_400_to_corrupt() {
    // 410 means the code timed out and a new one will fix it. 400 means the
    // stored secret is malformed, which retrying the same code never fixes --
    // "error" is reserved for exactly this case and nothing else.
    assert_eq!(poll_raising(NetError::http(410)).status, "expired");
    assert_eq!(poll_raising(NetError::http(400)).status, "error");
}

#[test]
fn test_poll_over_cap_redeem_reads_as_expired() {
    // The server answers an over-cap redeem with 410 too -- the player learns
    // the real reason (too many computers) on the website at claim time, so the
    // game just treats it like any other timed-out code.
    assert_eq!(poll_raising(NetError::http(410)).status, "expired");
}

#[test]
fn test_poll_maps_503_to_retry_not_error() {
    // A 5xx is the server's own trouble, not a verdict on this device_code --
    // unlike a 400, the same code is worth polling again on the next tick.
    assert_eq!(poll_raising(NetError::http(503)).status, "retry");
}

#[test]
fn test_poll_maps_other_http_statuses_to_retry() {
    // Nothing except 400 and 410 is meaningful here; nothing else should be
    // "error" (terminal) either -- nearly anything else the server could answer
    // with is worth polling again rather than sending the player back through
    // a whole new activation code.
    for code in [401u16, 403, 404, 429, 500, 502] {
        assert_eq!(poll_raising(NetError::http(code)).status, "retry");
    }
}

#[test]
fn test_poll_never_raises_on_network_trouble() {
    // Polling runs on a timer while the player waits; a transient blip must
    // not crash the menu, and must not be presented as the terminal "error"
    // that only a malformed device_code (HTTP 400) gets -- the next poll, a
    // few seconds later, may well succeed on its own.
    let result = poll_raising(NetError::other("OSError", "connection reset"));
    assert_eq!(result.status, "retry");
}

#[test]
fn test_poll_never_raises_on_malformed_200() {
    let result = poll_activation(
        &an_activation(),
        &replying(json!({"status": "some-unexpected-shape"})),
    );
    assert_eq!(result.status, "retry");
}

#[test]
fn test_poll_never_raises_on_a_null_200_body() {
    // Regression: reply.get(...) ran unguarded on whatever the transport
    // returned, so a 200 with a non-mapping body (None here) raised straight
    // into the caller instead of coming back as a retryable result.
    let result = poll_activation(&an_activation(), &replying(Value::Null));
    assert_eq!(result.status, "retry");
}

#[test]
fn test_poll_never_raises_on_a_list_200_body() {
    let result = poll_activation(&an_activation(), &replying(json!(["not", "a", "mapping"])));
    assert_eq!(result.status, "retry");
}

#[test]
fn test_poll_never_raises_on_a_dict_missing_status() {
    let result = poll_activation(
        &an_activation(),
        &replying(json!({"driver_id": "rig-hauler"})),
    );
    assert_eq!(result.status, "retry");
}

#[test]
fn test_base_url_reused_from_online_presence() {
    // online_activation must not fork its own base_url -- one FREIGHT_FATE_ONLINE_URL
    // override has to redirect every Orinks endpoint the game talks to.
    assert_eq!(base_url(), freight_fate::online_presence::base_url());
}

#[test]
fn test_machine_key_is_stable_within_a_process() {
    let dir = data_dir();
    let first = machine_key(dir.path());
    let second = machine_key(dir.path());
    assert_eq!(first, second);
    assert_eq!(first.len(), 32);
}
